examples/fabric/byot/trainer.py

```diff
@@ -0,0 +1,409 @@
+
+            # only increase global step if optimizer stepped
+            self.global_step += int(should_optim_step)
+
+            # stopping criterion on step level
+            if self.max_steps is not None and self.global_step >= self.max_steps:
+                self.should_stop = True
+                break
+
+        self.fabric.call("on_train_epoch_end")
+
+    def val_loop(
+        self, model: LightningModule, val_loader: Optional[DataLoader], limit_batches: Union[int, float] = float("inf")
+    ):
+        # no validation if val_loader wasn't passed
+        if val_loader is None:
+            return
+
+        # no validation but warning if val_loader was passed, but validation_step not implemented
+        elif val_loader is not None and not is_overridden("validation_step", _unwrap_objects(model), LightningModule):
+            rank_zero_warn(
+                "Your LightningModule does not have a validation_step implemented, "
+                "but you passed a validation dataloder. Skipping Validation."
+            )
+            return
+
+        self.fabric.call("on_validation_model_eval")  # calls `model.eval()`
+
+        # TODO: inference mode?
+        torch.set_grad_enabled(False)
+
+        self.fabric.call("on_validation_epoch_start")
+
+        iterable = self.pbar_wrapper(
+            val_loader,
+            total=min(len(val_loader), limit_batches),
+            desc="Validation Sanity Check" if self.global_step == 0 else "Validation",
+        )
+
+        for batch_idx, batch in enumerate(iterable):
+
+            # end epoch if stopping training completely or max batches for this epoch reached
+            if self.should_stop or batch_idx >= limit_batches:
+                self.fabric.call("on_validation_epoch_end")
+                return
+
+            self.fabric.call("on_validation_batch_start", batch, batch_idx)
+
+            out = model.validation_step(batch, batch_idx)
+            # avoid gradients in stored/accumulated values -> prevents potential OOM
+            out = apply_to_collection(out, torch.Tensor, lambda x: x.detach())
+
+            self.fabric.call("on_validation_batch_end", out, batch, batch_idx)
+            self._current_val_return = out
+
+            self._format_iterable(iterable, self._current_val_return, "val")
+
+        self.fabric.call("on_validation_epoch_end")
+
+        self.fabric.call("on_validation_model_train")
+        # TODO: Inference mode?
+        torch.set_grad_enabled(True)
+
+    def training_step(self, model, optimizer, batch, batch_idx):
+        outputs: Union[torch.Tensor, Mapping[str, Any]] = model.training_step(batch, batch_idx=batch_idx)
+
+        loss = outputs if isinstance(outputs, torch.Tensor) else outputs["loss"]
+
+        self.fabric.call("on_before_backward", loss)
+        self.fabric.backward(loss)
+        self.fabric.call("on_after_backward")
+        # TODO: reroute configure_gradient_clipping in LM to fabric.gradient clipping
```

Comment:
should be relatively easy, I'll prepare a PR for this after which we'll then be able to add this here.

Reply:
I wouldnt implement clipping. This is just an example that is meant to be forked. Anything that's not key in driving the training loop I wouldnt include. Keep it simple!

Reply:
I'd keep it simple and just call the minimal set of hooks to have things work. Otherwise we're going again into the direction of replacing the Trainer

Reply:
I'll remove it!